salt/loader_core.py

```diff
@@ -0,0 +1,941 @@
+# -*- coding: utf-8 -*-
+'''The Salt loader is the core to Salt's plugin system, the loader scans
+directories for python loadable code and organizes the code into the
+plugin interfaces used by Salt.
+
+Core functions are contained here and should only be used through
+salt.loader.
+'''
+
+# Import python libs
+from __future__ import absolute_import
+import os
+import imp
+import sys
+import time
+import logging
+import inspect
+import tempfile
+import functools
+from collections import MutableMapping
+from zipimport import zipimporter
+
+# Import 3rd-party libs
+try:
+    import pkg_resources
+    HAS_PKG_RESOURCES = True
+except ImportError:
+    HAS_PKG_RESOURCES = False
+import salt.ext.six as six
+
+# Import salt libs
+from salt.utils.decorators import Depends
+import salt.utils.context
+import salt.utils.lazy
+import salt.utils.event
+import salt.utils.odict
+
+# Solve the Chicken and egg problem where grains need to run before any
+# of the modules are loaded and are generally available for any usage.
+import salt.modules.cmdmod
+
+__salt__ = {
+    'cmd.run': salt.modules.cmdmod._run_quiet
+}
+log = logging.getLogger(__name__)
+
+SALT_BASE_PATH = os.path.abspath(os.path.dirname(salt.__file__))
+LOADED_BASE_NAME = 'salt.loaded'
+
+if six.PY3:
+    # pylint: disable=no-member,no-name-in-module,import-error
+    import importlib.machinery
+    SUFFIXES = []
+    for suffix in importlib.machinery.EXTENSION_SUFFIXES:
+        SUFFIXES.append((suffix, 'rb', 3))
+    for suffix in importlib.machinery.BYTECODE_SUFFIXES:
+        SUFFIXES.append((suffix, 'rb', 2))
+    for suffix in importlib.machinery.SOURCE_SUFFIXES:
+        SUFFIXES.append((suffix, 'r', 1))
+    # pylint: enable=no-member,no-name-in-module,import-error
+else:
+    SUFFIXES = imp.get_suffixes()
+
+# Will be set to pyximport module at runtime if cython is enabled in config.
+pyximport = None
+
+
+def static_loader(
+        opts,
+        ext_type,
+        tag,
+        pack=None,
+        int_type=None,
+        ext_dirs=True,
+        ext_type_dirs=None,
+        base_path=None,
+        filter_name=None,
+        ):
+    funcs = LazyLoader(
+        module_dirs(
+            opts,
+            ext_type,
+            tag,
+            int_type,
+            ext_dirs,
+            ext_type_dirs,
+            base_path,
+        ),
+        opts,
+        tag=tag,
+        pack=pack,
+    )
+    ret = {}
+    funcs._load_all()
+    if filter_name:
+        funcs = FilterDictWrapper(funcs, filter_name)
+    for key in funcs:
+        ret[key] = funcs[key]
+    return ret
+
+
+def module_dirs(
```

Comment:
I renamed this from ``_module_dirs`` to ``module_dirs`` to squelch pylint warnings - it also is used outside of the loader (in a few spots) and consequently the ``_`` prefix is questionable.